### PR TITLE
Add geothermal electricity extraction support

### PR DESCRIPTION
**Pull request by bpulluta**
## Overview

This PR adds geothermal electricity as a supported extraction technology in COMPASS, including the extraction schema and plugin configuration needed to discover, retrieve, and extract structured ordinance data from jurisdictions governing utility-scale geothermal electricity generation.

**Additional improvements:**
- Clarifies the output directory policy for the CLI, ensuring users understand that `--out_dir_exists` must be set via the CLI
- Fixes two bugs in the retrieval layer affecting all technologies.

---

### New: Geothermal Electricity Extraction

**Files added:**
- `compass/extraction/geothermal_electricity/geothermal_schema.json` — Defines 29 extractable features including setbacks, permitting, noise limits, zoning classifications, decommissioning, and drilling requirements.
- `compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml` — Configures search queries, website scoring keywords, heuristic filters, and document collection behavior tuned for geothermal electricity ordinances.

The schema follows the standard COMPASS one-shot extraction format and is compatible with the existing `compass process` pipeline.

---

### CLI Improvements

- The `--out_dir_exists` option for handling existing output directories is now available for users to overwrite, fail, or create a new directory automatically
- Help text and logic ensure robust, user-friendly, and safe behavior for both interactive and automated runs.

---

### Bug Fixes

**Bug Fix 1 — PDF URLs with spaces failed to download**
- `crawl4ai` can return document URLs with raw spaces in the path. These are now percent-encoded before download, ensuring correct retrieval.
- File: `compass/scripts/download.py`

**Bug Fix 2 — Anchor text was never used in link scoring**
- Anchor text is now correctly populated and used in link scoring, improving document discovery.
- File: `compass/web/website_crawl.py`

### CODE REVIEW

File: compass/_cli/process.py [hunk 1]
```diff
@@ -49,10 +55,30 @@
...
+def process(config, verbose, no_progress, plugin, out_dir_exists):
     """Download and extract ordinances for a list of jurisdictions"""
     config = load_config(config)
 
+    if out_dir_exists is not None:
+        out_dir_policy = out_dir_exists
+    elif sys.stdin.isatty():
+        out_dir_policy = "prompt"
+    else:
+        out_dir_policy = "fail"
```
Comment (ppinchuk) [inline compass/_cli/process.py]: Please move this to be inside of `_resolve_out_dir_conflict`. This helps keep things organized

File: compass/_cli/process.py [hunk 2]
```diff
@@ -128,3 +154,80 @@ def _setup_cli_logging(console, verbosity_level, log_level="INFO"):
...
+def _resolve_out_dir_conflict(out_dir, policy):
+    """Handle existing output directory using the selected policy"""
+    out_dir = Path(out_dir)
+    policy = policy.lower()
+
+    if not out_dir.exists():
+        return out_dir
+
+    if policy == "fail":
+        return out_dir
```
Comment (ppinchuk) [inline compass/_cli/process.py]: please combine these into `if not out_dir.exists() or policy == "fail"`

File: compass/_cli/process.py [hunk 3]
```diff
@@ -128,3 +154,80 @@ def _setup_cli_logging(console, verbosity_level, log_level="INFO"):
...
+        if overwrite:
+            click.echo(f"Overwriting existing output directory: {out_dir!s}")
+            shutil.rmtree(out_dir)
+            return out_dir
+
+        msg = (
+            "Run cancelled. Please update out_dir in config, or rerun with "
+            "--out_dir_exists increment/overwrite."
+        )
+        raise click.ClickException(msg)
```
Comment (ppinchuk) [inline compass/_cli/process.py]: please move this logic into a separate function for code clarity

File: compass/_cli/process.py [hunk 4]
```diff
@@ -128,3 +154,80 @@ def _setup_cli_logging(console, verbosity_level, log_level="INFO"):
...
+    """
+    Create the next available output directory suffix with
+    versioning
+    """
+    idx = 2
+    while True:
+        candidate = out_dir.parent / f"{out_dir.name}_v{idx}"
+        if not candidate.exists():
+            return candidate
+        idx += 1
```
Comment (ppinchuk) [inline compass/_cli/process.py]: Please add a guard (e.g. if index goes over 1k or even 1M) to prevent infinite loops

File: compass/plugin/one_shot/components.py [hunk 5]
```diff
@@ -296,11 +296,18 @@ async def _process(self, text_chunks):
...
+_MAGIC_MINUTE_59 = 59
+
+
 class SchemaOrdinanceParser(SchemaOutputLLMCaller, BaseParser):
     """Base class for parsing structured data"""
 
     DATA_TYPE_SHORT_DESC = None
-    """Optional short description of the type of data being extracted
+    """
+    Optional short description of the type of data being extracted
```
Comment (ppinchuk) [inline compass/plugin/one_shot/components.py:303]: Please undo this change. The style of this repo is to start a docstring immediately after the triple quote

File: compass/plugin/one_shot/components.py [hunk 6]
```diff
@@ -413,3 +422,108 @@ def _to_dataframe(self, data):
...
+                    else hour + _MAGIC_HOUR_12
+                )
+            out.append(f"{hour:02d}:{minute:02d}")
+        out.extend(
+            [
+                f"{int(match.group(1)):02d}:{int(match.group(2)):02d}"
+                for match in hhmm_pattern.finditer(text)
+            ]
+        )
+        return sorted(set(out))
```
Comment (ppinchuk) [inline compass/plugin/one_shot/components.py]: Please move this to a regular module-level function instead of a static method

File: compass/plugin/one_shot/components.py [hunk 7]
```diff
@@ -296,11 +296,18 @@ async def _process(self, text_chunks):
         return text_summary
 
 
+# Constants for magic values
+_MAGIC_NEIGHBOR_CHUNK_COUNT = 2
+_MAGIC_HOUR_12 = 12
+_MAGIC_MINUTE_59 = 59
```
Comment (ppinchuk) [inline compass/plugin/one_shot/components.py]: I think we can be better about the names here

File: compass/plugin/one_shot/components.py [hunk 8]
```diff
@@ -413,3 +422,108 @@ def _to_dataframe(self, data):
...
+
+    def _pp_bounded_time_from_summary(self, row, step):
+        """
+        Prefer bounded time windows from summary over fallback
+        values
+        """
+        feature = (row.get("feature") or "").casefold()
+        source_field = step.get("source_field", "summary")
+        source_text = row.get(source_field) or ""
+        time_values = self._extract_times_from_text(source_text)
```
Comment (ppinchuk) [inline compass/plugin/one_shot/components.py]: I don't love the idea of doing manual processing on the summary text instead of relying on the LLM output. The summary might have multiple pairs of start/end times that need to be interpreted using the text context instead of just heuristic guessing. 

Can you explain why you chose not to go with the LLM outputs here?

File: compass/web/url_utils.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+from urllib.parse import quote, urlsplit, urlunsplit
+
+
+def _sanitize_url(url):
+    """Encode unsafe URL characters while preserving URL semantics"""
+    parsed = urlsplit(url)
+    path = quote(parsed.path, safe="/:@-._~!$&'()*+,;=")
+    query = quote(parsed.query, safe="=&;%:@-._~!$&'()*+,;/?:")
+    fragment = quote(parsed.fragment, safe="")
+    return urlunsplit((parsed.scheme, parsed.netloc, path, query, fragment))
```
Comment (ppinchuk) [inline compass/web/url_utils.py]: I like that you moved this to be it's own utility function, but please make it public (no leading underscore) and document it properly

File: compass/web/website_crawl.py [hunk 10]
```diff
@@ -532,6 +499,7 @@ def _extract_links_from_html(text, base_url):
     return {
         _Link(
             title=title,
+            text=title,
```
Comment (ppinchuk) [inline compass/web/website_crawl.py]: Please remove this change. The title is already given in the link - there should be no reason to duplicate it as text, right?

**Review by ppinchuk — CHANGES_REQUESTED** [review]
I didn't dive deep into the schema yet, but there are a few code things I'd like to address before focusing on the schema.

I _really_ like the idea of adding a configurable post-processing pipeline to the schema. I envision that we could build up a "library" of post-processing functions that are generic enough where different schemas can use them. Very cool idea and we should definitely use it. 

I'm quite concerned about parsing the summary for drilling hour windows instead of relying on the LLM output. Are you sure this is required? Can you explain why you went with that choice @bpulluta? Is there not a way to shape the output of the LLM instead? I'm worried about the case where the summary has multiple windows, which we already saw in the sample documents you had

**Comment by ppinchuk** [comment]
> I didn't dive deep into the schema yet, but there are a few code things I'd like to address before focusing on the schema.
> 
> I _really_ like the idea of adding a configurable post-processing pipeline to the schema. I envision that we could build up a "library" of post-processing functions that are generic enough where different schemas can use them. Very cool idea and we should definitely use it.
> 
> I'm quite concerned about parsing the summary for drilling hour windows instead of relying on the LLM output. Are you sure this is required? Can you explain why you went with that choice @bpulluta? Is there not a way to shape the output of the LLM instead? I'm worried about the case where the summary has multiple windows, which we already saw in the sample documents you had

OK on second look, it looks like the summary is only being used as a fallback. That might be ok actually. I think you can disregard the comments about the summary parsing

File: compass/plugin/one_shot/components.py (hunk 6 above)
Comment (ppinchuk) [reply compass/plugin/one_shot/components.py]: And once you do, could you please add some tests for this function specifically? I think it's critical to ensure that it does what we expect it to do. Try to think of typical edge cases we might see in the LLM output for this field and use those for testing

File: tox.ini
Comment (castelao) [inline tox.ini:1]: I'm confused here. The latest version of crawl4ai is 0.8.6, isn't it? Is that a typo and you intended `0.7.2`?

Independent of the version used, I suggest having the latest with no constraints. Any constraint to allow `latest `to run, should actually go in the definition of the project in the `pyproject.toml` to guarantee that people not using `pixi` can run the package as well. Sorry, I'm just viewing this PR. I merged another PR early today that might address what was the intention here.

File: compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+cache_llm_generated_content: true
+
+query_templates:
+    - "filetype:pdf {jurisdiction} geothermal power plant ordinance"
+    - "geothermal electricity generation ordinance {jurisdiction}"
+    - "{jurisdiction} geothermal energy facility zoning ordinance"
+    - "{jurisdiction} geothermal power plant land use code"
+    - "{jurisdiction} geothermal code of ordinances"
+    - "{jurisdiction} geothermal conditional use permit"
+    - "{jurisdiction} geothermal special use permit"
```
Comment (ppinchuk) [inline compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml]: Have you found these to be useful? I imagine if a permit application is found, it would just eventually get filtered out by our content filter, which seemingly makes these unused?

File: compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+    - "{jurisdiction} geothermal drilling permit regulations"
+    - "{jurisdiction} geothermal resource development statute"
+    - "Where can I find the legal text for geothermal power plant zoning ordinances in {jurisdiction}?"
+    - "What is the specific legal information regarding zoning ordinances for geothermal electricity generation facilities in {jurisdiction}?"
```
Comment (ppinchuk) [inline compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml]: Keep in mind that the default number of docs that will be parsed is 5. If all your queries return a different URL, then the results from some of these will not be considered. It might be worthwhile to reduce these to the top 5 most important queries so that we don't eat up our search quota as fast, unless you have found that each one of these uniquely returns a document you are interested in.

File: compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+    - "geothermal reservoir"
+    - "geothermal permit"
+    - "geothermal ordinance"
+    - "geothermal zoning"
+    - "code of ordinances"
+    - "land use code"
+    - "use table"
+    - "zoning ordinance"
+    - "special use permit"
+    - "conditional use permit"
```
Comment (ppinchuk) [inline compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml]: Couple of notes here:

- Most of these should be under `good_tech_phrases`. The reason for that is that keywords are a direct lookup, while phrases use ngram comparisons. The latter is more robust when you have spaces, since your phrase may be broken up by things like newlines (e.g. `geothermal\nwell`). I would move anything with a space into `good_tech_phrases` and keep `good_tech_keywords` restricted to single words. 

- Keep in mind that the way the heuristic works is that it checks for matches across `good_tech_keywords`, `good_tech_acronyms`, and `good_tech_phrases`. The matches across all three categories are totaled, and if you find at least two matches (by default), the document passes the heuristic check. Here is what that means in practice:
    - If you repeat words or phrases across categories, then a single match will actually count for two or more, which would automatically qualify the document.
    - If you keep more "generic" terms like "land use code", "use table", "zoning ordinance", etc, your heuristic will pass through non-geothermal ordinance documents.
  
   Both of these basically mean your heuristic becomes much less effective. While it's ok (and maybe even preferred) for the heuristic to be permissive, if it just lets most kinds of documents through, you will still have to do a lot of filtering afterwards, and in the case of webcrawling, you might miss the correct document entirely. 


My suggestion for both of these is to keep the heuristic as focused as possible on geothermal electricity. You can kind of see how I did this for GHP's here: https://github.com/NatLabRockies/COMPASS/blob/92e7756fee18c4380cf99a3587cc2106d7c0e712/compass/extraction/ghp/plugin_config.yaml#L26-L91

File: compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+    - "rotor diameter"
+    - "tip height"
+    - "nacelle"
+    - "battery storage"
+    - "energy storage system"
+    - "hydroelectric"
+    - "biomass"
+    - "cannabis"
+    - "cannabis cultivation"
+    - "commercial cannabis"
```
Comment (ppinchuk) [inline compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml]: Please add 

`collection_prompts: True`

at the bottom of this config. Without it, you will always pass the full document into the extraction context, which could explain why there were some rate limit issues as well as why the API cost for the ~200 counties was so high.

File: compass/extraction/geothermal_electricity/geothermal_schema.json [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+        "prohibited use districts": {
+          "description": "Extract all districts, zones, overlays, or similar land-use categories where geothermal electricity facilities are explicitly and unconditionally prohibited. Use an array of district names in value and quote the prohibitory language in summary. Do not include districts where geothermal electricity may still be allowed with a permit, variance, overlay, or other conditional path."
+        }
+      }
+    },
+    "prohibition_features": {
+      "description": "Prohibitions, bans, or moratoria on geothermal electricity exploration, drilling, facility siting, or deployment.",
+      "properties": {
+        "prohibitions": {
+          "description": "Extract currently effective bans, moratoria, or explicit prohibitions on geothermal electricity exploration, drilling, well development, plant construction, or facility siting. Include fracking or hydraulic-fracturing bans only when the ordinance explicitly uses them to ban, limit, or condition geothermal electricity activity. If there are carve-outs, exceptions, or conditional permitting paths that still allow the project, do not classify the rule as a prohibition."
```
Comment (ppinchuk) [inline compass/extraction/geothermal_electricity/geothermal_schema.json]: Let's add something like "For moratoria, include start/end timing details in summary; if the ordinance states an end date that has already passed, omit this feature."

The idea is that we want details about the date range for the moratorium, and we don't want to report it if it's expired.

File: compass/extraction/geothermal_electricity/geothermal_schema.json [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+    "fencing",
+    "color requirements",
+    "lighting requirements",
+    "visual impact assessment",
+    "seismic monitoring plan",
+    "bond requirement",
+    "decommissioning",
+    "prohibitions"
+  ]
+}
```
Comment (ppinchuk) [inline compass/extraction/geothermal_electricity/geothermal_schema.json:394]: Nice, thank you for matching the structure of the GHP schema! It's nice to have some cohesion 😄

**Review by ppinchuk — CHANGES_REQUESTED** [review]
A few more comments specifically for the plugin and schema configs. Excited to get this merged in!

**Comment by castelao** [comment]
How do we move this one forward? @ppinchuk 's comments are relevant and should be addressed or it would be best to discuss why not. @bpulluta , would you like help to address the open issues on this PR?

As we rely more and more on the one-shot approach, properly closing this PR is important to set the tone for the other technologies on the way. That will save us a lot of time down the road.

File: compass/plugin/one_shot/components.py (hunk 8 above)
Comment (bpulluta) [reply compass/plugin/one_shot/components.py]: Implemented as requested: I removed the bounded-time summary postprocess from SchemaOrdinanceParser so drilling times now rely entirely on LLM outputs.

File: compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml (hunk 11 above)
Comment (bpulluta) [reply compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml]: @ppinchuk I haven't specifically tested if these by themselves improve the data collection, I'll try removing these and see if that affects the permits found

File: compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml (hunk 12 above)
Comment (bpulluta) [reply compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml]: @ppinchuk same as above, i'm going to try to test it by reducing to 5 and i'll let you know if it changes anything

File: compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml (hunk 13 above)
Comment (bpulluta) [reply compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml]: @ppinchuk thanks for the example, same as before I'll make incremental changes to the YAML and ensure that it doesnt degrade the current out. My only concern is that making these changes might change the quality of the output but perhaps not. I'll do some testing and come back to these comments

File: compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml (hunk 11 above)
Comment (bpulluta) [reply compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml]: @ppinchuk  Tested sequentially on a couple of counties. I removed permit focused query templates first and reran extraction. In Cheyenne, the direct drilling permit application URL stopped appearing in the candidate set, while the same four county zoning documents still remained after jurisdiction filtering. Keeping this change because it reduced noise without hurting retained county legal docs on the subset.

File: compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml (hunk 12 above)
Comment (bpulluta) [reply compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml]: @ppinchuk  I then reduced query templates to five and reran the same county subset. Cheyenne still produced the same four retained county docs after jurisdiction filtering, and Chaffee still surfaced county land use sources in candidates. This indicates no material quality loss on the subset while using fewer queries.

File: compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml (hunk 13 above)
Comment (bpulluta) [reply compass/extraction/geothermal_electricity/geothermal_plugin_config.yaml]: @ppinchuk  For the heuristic cleanup, I moved multi word items into good_tech_phrases, kept single tokens in good_tech_keywords, and removed generic ordinance and zoning phrases. On the same subset, Cheyenne behavior remained stable with four retained county docs after jurisdiction filtering and irrelevant out of jurisdiction material rejected. I am keeping this change based on that non regression signal.

**Review by bpulluta — COMMENTED** [review]
reviewed and made the changes outlined in the comments. No regression was observed in reruning the geothermal electricity extraction

**Review by ppinchuk — APPROVED** [review]
Thanks again @bpulluta! Looks great!
